FWCore/Services/plugins/StallMonitor.cc

```diff
@@ -0,0 +1,403 @@
+        << "# postEventReadFromSource       " << step::postEventReadFromSource    << "   <Stream ID> <Module ID> <Time since beginJob (ms)>\n"
+        << "# postModuleEvent               " << step::postModuleEvent            << "   <Stream ID> <Module ID> <Time since beginJob (ms)>\n"
+        << "# postEvent                     " << step::postEvent                  << "   <Stream ID> <Run#> <LumiBlock#> <Event#> <Time since beginJob (ms)>\n";
+    file_.write(oss.str());
+  }
+}
+
+void StallMonitor::fillDescriptions(ConfigurationDescriptions& descriptions)
+{
+  ParameterSetDescription desc;
+  desc.addUntracked<std::string>("fileName", filename_default)->setComment("Name of file to which detailed timing information should be written.\n"
+                                                                           "An empty filename argument (the default) indicates that no extra\n"
+                                                                           "information will be written to a dedicated file, but only the summary\n"
+                                                                           "including stalling-modules information will be logged.");
+  desc.addUntracked<double>("stallThreshold", threshold_default)->setComment("Threshold (in seconds) used to classify modules as stalled.\n"
+                                                                             "Millisecond granularity allowed.");
+  descriptions.add("StallMonitor", desc);
+  descriptions.setComment("This service keeps track of various times in event-processing to determine which modules are stalling.");
+}
+
+void StallMonitor::preModuleConstruction(ModuleDescription const& md)
+{
+  // Module labels are dense, so if the module id is greater than the
+  // size of moduleLabels_, grow the vector to the correct index and
+  // assign the last entry to the desired label.  Note that with the
+  // current implementation, there is no module with ID '0'.  In
+  // principle, the module-information vectors are therefore each one
+  // entry too large.  However, since removing the entry at the front
+  // makes for awkward indexing later on, and since the sizes of these
+  // extra entries are on the order of bytes, we will leave them in
+  // and skip over them later when printing out summaries.  The
+  // extraneous entries can be identified by their module labels being
+  // empty.
+  auto const mid = md.id();
+  if (mid < moduleLabels_.size()) {
+    moduleLabels_[mid] = md.moduleLabel();
+  }
+  else {
+    moduleLabels_.resize(mid+1);
+    moduleLabels_.back() = md.moduleLabel();
+  }
+}
+
+void StallMonitor::postBeginJob()
+{
+  // Since a (push,emplace)_back cannot be called for a vector of a
+  // type containing atomics (like 'StallStatistics')--i.e. atomics
+  // have no copy/move-assignment operators, we must specify the size
+  // of the vector at construction time.
+  moduleStats_ = std::vector<StallStatistics>(moduleLabels_.size());
```

Comment:
For your case, you could hand write copy constructor for `StallStatistics` since the copy doesn't have to be thread-safe. I'm fine either way.